### DISCUSSION TITLE
Add JUnit RunListener to Tycho Surefire Plugin

### DISCUSSION THREAD

**Discussion by vaclavHala** [eclipse-tycho/tycho#1258] [Q&A]
We have some custom [RunListener](https://junit.org/junit4/javadoc/4.12/index.html?org/junit/runner/notification/RunListener.html) we'd like to add to each of our tests (i.e. register it in parent pom and have all test bundles inherit it).

So far I managed to get the listener working by registering it like this in the parent pom
```
<plugin>
    <groupId>org.eclipse.tycho</groupId>
    <artifactId>tycho-surefire-plugin</artifactId>
    <configuration>
        <providerProperties>
            <property>
                <name>listener</name>
                <value>my.custom.Listener</value>
            </property>
        </providerProperties>
```
and then adding bundle containing this class as explicit dependency to `META-INF/MANIFEST.MF` of every test bundle.

While this works, I'm not happy about having to add it to each test as every time we add a new test bundle the build breaks (failed to find the listener class).

So what I tried to do was mark the bundle containing the listener as `Fragment-Host: org.eclipse.tycho.surefire.osgibooter`, as that seems to be the bundle which does the lookup by reflection at runtime, and adding the listener bundle as extra requirement to `target-platform-configuration` in the parent. Problem with this approach is that target resolution fails with such setup:
```
[ERROR] Cannot resolve project dependencies:
[ERROR]   Software being installed: my.custom.listener 1.0.0
[ERROR]   Missing requirement: my.custom.listener 1.0.0 requires 'osgi.bundle; org.eclipse.tycho.surefire.osgibooter 0.0.0' but it could not be found
```
which seems wrong to me as the booter obviously is part of the final target, this can be seen by running any tests with the `-Dtycho.printBundles` flag.

Finally, my question - is the approach with fragment I took reasonable and should I create a feature request for exposing the booter in the target? Is there some better way I didn't think of to achieve what I'm trying to do with what is already available (i.e. add some code visible to tycho-surefire classpath without having to explicitly depend on it from the test bundle)?

I have found thread asking for pretty much the same thing https://www.eclipse.org/lists/tycho-user/msg06256.html but no resolution there.

**Comment by laeubi** [comment]
> which seems wrong to me as the booter obviously is part of the final target

You are mixing thinks here, the booter is not part of the project resolution (e.g. you will never really see it on the compile path) but only part of the test-execution framework, that are different things.

> and adding the listener bundle as extra requirement to `target-platform-configuration` in the parent

I think that https://www.eclipse.org/tycho/sitedocs/tycho-surefire-plugin/integration-test-mojo.html#dependencies is more suitable for this (never checked if a plain jar is then even enough).

**Reply by vaclavHala** [reply]
My problem is not adding the jar into the target, I did that by the `target-platform-configuration` and it seems to work just fine.

The problem is that when this code eventually gets called https://github.com/apache/maven-surefire/blob/master/surefire-providers/common-junit4/src/main/java/org/apache/maven/surefire/common/junit4/JUnit4RunListenerFactory.java#L36 even though my bundle is present in the target, it is not visible to the provider bundle doing the reflection lookup (I may be wrong about the booter, maybe it is `org.eclipse.tycho.surefire.junit4` now that I'm looking at the stack traces again). That led me to try the fragment approach as that is I think the "correct OSGI way" of adding some resource to existing bundle without modifying the code of the existing bundle itself.

> You are mixing thinks here, the booter is not part of the project resolution (e.g. you will never really see it on the compile path) but only part of the test-execution framework, that are different things.

Ok, would there be some way for me to add my bundle into this execution framework then? As user extension/plugin I mean. That should be what I'm trying to do I think

>  you will never really see it on the compile path

I read this part again and I apparently failed to mention I worked around this by adding the booter as `provided` dependency to the pom of my listener. So the compile step passes just fine and I have a packaged jar. Problem appears at runtime when I add this jar to my runtime.

Also I should add that we are using both `default` and `p2Installed` runtime modes. This option https://www.eclipse.org/tycho/sitedocs/tycho-surefire-plugin/test-mojo.html#frameworkExtensions seemed promising but doc says it cannot be used in `p2Installed` mode.

**Reply by laeubi** [reply]
> My problem is not adding the jar into the target, I did that by the `target-platform-configuration` and it seems to work just fine.

Obviously not, or you won't ask here

> Ok, would there be some way for me to add my bundle into this execution framework then? As user extension/plugin I mean.

That's what https://www.eclipse.org/tycho/sitedocs/tycho-surefire-plugin/integration-test-mojo.html#dependencies is for but as noted in the doc, this is a bit of a gray area with less coverage and seldom used.

**Reply by vaclavHala** [reply]
> Obviously not, or you won't ask here

Hm I don't agree with you there. I tried using the `<dependencies>` you suggest too but the result was exactly the same as using the `target-platform-configuration` - the plugin is present in the test runtime (listed when given the `-Dtycho.printRuntime`, can be found at runtime from my code via `BundleContext#getBundles`  ), but the `Listener` class cannot be found by reflection by the junit provider.

Now don't get me wrong, I'm not saying that all I have works, I'm just saying that I think there are several orthogonal parts that all need to work together.

One of the parts is providing list of bundles to run, this I think works for me.

Another part is making sure visibility between bundles present in the runtime is such that the junit provider can find my code. As I understand OSGI, just because two bundles are present in the runtime does not automatically mean they can access each other (using reflection) as is the case in standard java. Either plugin has to declare in manifest it wants to see some other plugin, or the other plugin has to declare itself as fragment for the first (what I tried to do). 

Is my understanding of this wrong?

**Reply by laeubi** [reply]
I think the best would be to provide a small example to demonstrate the issue, its generally easier to discuss on the code. Beside that, you should find out what bundle is really requesting the Listener, e.g. is it your test-probe? is it some Tycho code? Or do surefire itself request this and what classloader is used.

**Reply by vaclavHala** [reply]
sure, here it is https://github.com/vaclavHala/tycho-surefire-target-demo

I hope the example will make it clear what I'm trying to do. Please let me know if there is still something I should explain better

**Reply by laeubi** [reply]
The example states that removing "Fragment-Host" makes it fail, why do you want to remove this? It would probably also work to attatch to the junit bundles itself if you think that is more sufficient.

**Reply by vaclavHala** [reply]
I don't want to remove it, I think it needs to be there to make the Listener visible, it "fails" both with and without the fragment host
* with the fragment-host it fails to resolve target and start at all
* without the fragment-host it fails at runtime with class not found exception thrown from the junit provider

The comment about removing the fragment-host was meant to show why I think it needs to be there as without it you can see the listener bundle is present in the target runtime but cant be found by the provider

I tried using `org.eclipse.tycho.surefire.junit4` as the fragment host but that fails too because this itself is a fragment of the `org.eclipse.tycho.surefire.osgibooter` and I don't think fragments can have fragments of their own. Or did you mean some other bundle when you said `the junit bundles itself`?

**Reply by laeubi** [reply]
> I don't want to remove it, I think it needs to be there to make the Listener visible

It depends a bit, e.g one might use `Dynamic-Import` as well.

> or did you mean some other bundle when you said `the junit bundles itself`?

for example `org.junit` bundle.

>  it "fails" both with and without the fragment host

Didn't you say it has worked before?

**Reply by vaclavHala** [reply]
> It depends a bit, e.g one might use Dynamic-Import as well.

But that would require me to modify the tycho/junit bundles which I don't want to do as the `Dynamic-Import` has to be given in the bundle "requiring" something no? I only have control over the bundle "providing" something (my listener bundle).

> Didn't you say it has worked before?

I said adding the bundle to the test runtime works, that still stands yes. But the class from the bundle fails to be loaded via reflection by the booter. So the whole thing together does not work and never worker, which is why I'm asking here.


My ultimate goal is to run the code in the example I provided and `Hello from listener` gets printed. There are following additional constraints:
* I don't want to modify any of the junit/tycho/surefire code (such as adding dynamic import into the junit provider)
* The listener is not explicitly required by every test bundle (in my example there is just `my.test.a` but in reality there would be dozens of test bundles and new ones are frequently added), it is just added to the test runtime via target configuration and registered by provider property once in the parent pom

**Reply by vaclavHala** [reply]
Got an update on this.

I tried changing the fragment-host to `org.junit` but that did not help. I think because the booter bundle depends on junit via package import rather than whole bundle import the code I add by my fragment is still not visible to the ProviderFactory.

However looking at the manifest of the booter it requires `org.eclipse.core.runtime` by bundle import so attaching my fragment to that and explicitly exporting package with my Listener class (so it gets exposed to booter along with other exports from the core.runtime bundle) does the trick. I updated the example code with this change to show it works.

I feel what I did here is more of a hack than proper solution since it depends on implementation of the booter (its imports in manifest). Using the booter as the fragment host would be more correct solution as it is the intuitive thing to do (in my opinion). So I still think it would be reasonable to have the option to use it as fragment host as I originally described. What do you think?

**Reply by laeubi** [reply]
All of this relies on internal implementation details and might break in a future Tycho release. "global listeners" is just something that do not work well in OSGi and the standard way would be to register the listener programmatically.

**Reply by vaclavHala** [reply]
> the standard way would be to register the listener programmatically

Could you elaborate on this please? What would that look like?

**Reply by laeubi** [reply]
JUnit 4: https://www.swtestacademy.com/junit-listeners/
JUnit 5: https://stackoverflow.com/questions/48595037/whats-the-junit-5-way-for-runlisteners

**Reply by vaclavHala** [reply]
Hm the junit4 approach seems to require me to register it in each test class explicitly which I wanted to avoid.

The junit5 approach with service loader seems promising, Ill look into that some more. Though I fear I will hit osgi visibility problem there as well as the surefire junit5 provider does not seem to be an osgi consumer of service loader services as described here http://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.loader.html . I might be wrong about that so I'll just try and see.

In either case, I guess we can close this. Thanks for your help
